### PR TITLE
Combine DSMB to FinalBlock

### PR DESCRIPTION
**Pull request by kaikawaliu**
## Description

This PR is to merge the DS microblock (Used to check txns sent to ds) to the finalblock consensus. Which can save one round of consensus during each epoch, shorten the inter block interval time.
So the basic idea is to put the composing of DS microblock will be integrated into the FinalBlock Consensus.

After the implementation, there are also five main test cases been simulated, which can break the consensus in different ways. This helped a lot in digging out potential risks that may hang the protocol, including: 
- [x] 1. Objective: Check fetching missing txn
   Scenario : DS leader has some txn that one of the backups doesn't have
   Adoption : Letting one of the backups accept less txns from lookup comparing to the others

- [x] 2. Objective: Check View Change due to dsblock check failure within FinalBlock consensus
   Scenario : DS leader has some txn that all of backups don't have
   Adoption : Letting all of the backups accept less txns from lookup comparing to the leader

- [X] 3. Objective: Check fetching missing microblock
   Scenario : DS leader has more microblock received than one of the backups
   Adoption : Letting one of the backups refuse some Microblock submission

- [X] 4. Objective: Check View Change after fetching missing microblock
   Scenario : DS leader has more microblock received than all of the backups
   Adoption : Letting all of the backups refuse some Microblock submission

- [X] 5. Objective: Check View Change due to TxBlock check failure within FinalBlock consensus
   Scenario : DS leader composed invalid TxBlock
   Adoption : Done by composing wrong state root hash

### Other features involved in this PR:
- Update processed transactions and remove processed transaction from created transaction - only when consensus was done
- replace the std::unordered_map of TxnHash with std::map TxnPool::GasIndex to ensure the consistency of transaction picking
- Attach the state delta for each microblock when the DS backup has missing and trying to fetch from leader. As a result, the backup can process transactions proposed by the DS leader during the ds microblock.
- Fixed a small issue with contract deployment

## Review Suggestion

Please feel free to approach me if had anything unclear.
### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [x] local machine test

- [ ] small-scale cloud test

### CODE REVIEW

File: src/libData/AccountData/TxnPool.h [hunk 1]
```diff
@@ -50,10 +49,21 @@ struct TxnPool {
...
+
   bool exist(const TxnHash& th) {
     return HashIndex.find(th) != HashIndex.end();
   }
 
+  bool get(const TxnHash& th, Transaction& t) {
+    if (!exist(th)) {
+      return false;
+    }
+    t = HashIndex[th];
```
Comment (bb111189) [inline src/libData/AccountData/TxnPool.h]: use .at()

File: src/libDirectoryService/FinalBlockPostProcessing.cpp [hunk 2]
```diff
@@ -386,26 +387,28 @@ void DirectoryService::ProcessFinalBlockConsensusWhenDone() {
...
+        // auto func2 = [this]() mutable -> void {
+        //   if (!m_stopRecvNewMBSubmission) {
+        m_stopRecvNewMBSubmission = true;
+        // m_mediator.m_node->RunConsensusOnMicroBlock();
+        //   }
+        // };
+
+        // DetachedFunction(1, func2);
+
+        // std::unique_lock<std::mutex>
```
Comment (bb111189) [inline src/libDirectoryService/FinalBlockPostProcessing.cpp]: can remove this whole chuck of comments?

File: src/libDirectoryService/MicroBlockProcessing.cpp [hunk 3]
```diff
@@ -451,22 +461,29 @@ bool DirectoryService::ProcessMissingMicroblockSubmission(
...
+                  "MicroBlocks fetched from leader have different size with "
+                  "stateDeltas");
+      return false;
+    }
+
+    for (unsigned int i = 0; i < microBlocks.size(); ++i) {
       if (!m_mediator.CheckWhetherBlockIsLatest(
-              microBlock.GetHeader().GetDSBlockNum() + 1,
-              microBlock.GetHeader().GetEpochNum())) {
+              microBlocks[i].GetHeader().GetDSBlockNum() + 1,
```
Comment (bb111189) [inline src/libDirectoryService/MicroBlockProcessing.cpp]: use .at()

File: src/libDirectoryService/MicroBlockProcessing.cpp [hunk 4]
```diff
@@ -451,22 +461,29 @@ bool DirectoryService::ProcessMissingMicroblockSubmission(
...
+              microBlocks[i].GetHeader().GetEpochNum())) {
```
Comment (bb111189) [inline src/libDirectoryService/MicroBlockProcessing.cpp]: use .at()

File: src/libDirectoryService/MicroBlockProcessing.cpp [hunk 5]
```diff
@@ -451,22 +461,29 @@ bool DirectoryService::ProcessMissingMicroblockSubmission(
...
         LOG_GENERAL(WARNING,
-                    "ProcessMicroblockSubmissionFromShardCore "
+                    "ProcessMissingMicroblockSubmission "
                     "CheckWhetherBlockIsLatest failed");
         return false;
       }
 
-      uint32_t shardId = microBlock.GetHeader().GetShardId();
+      uint32_t shardId = microBlocks[i].GetHeader().GetShardId();
```
Comment (bb111189) [inline src/libDirectoryService/MicroBlockProcessing.cpp]: use .at() and same for the rest

File: src/libData/AccountData/AccountStore.cpp [hunk 6]
```diff
@@ -291,10 +291,12 @@ void AccountStore::RevertCommitTemp() {
 
   // Revert changed
   for (auto const entry : m_addressToAccountRevChanged) {
+    // LOG_GENERAL(INFO, "Revert changed address: " << entry.first);
```
Comment (ansnunez) [inline src/libData/AccountData/AccountStore.cpp]: Do you want to keep this commented-out log message?

File: src/libData/AccountData/AccountStore.cpp [hunk 7]
```diff
@@ -291,10 +291,12 @@ void AccountStore::RevertCommitTemp() {
...
     (*m_addressToAccount)[entry.first] = entry.second;
     UpdateStateTrie(entry.first, entry.second);
   }
   for (auto const entry : m_addressToAccountRevCreated) {
+    // LOG_GENERAL(INFO, "Remove created address: " << entry.first);
```
Comment (ansnunez) [inline src/libData/AccountData/AccountStore.cpp]: Do you want to keep this commented-out log message?

Comment (kaikawaliu) [reply src/libData/AccountData/AccountStore.cpp]: Yeah, can just leave them there, after the state delta being fixed may also need to enable in local test and check again

File: src/libData/AccountData/AccountStore.cpp (hunk 6 above)
Comment (kaikawaliu) [reply src/libData/AccountData/AccountStore.cpp]: same as the below one

**Pull request merged**